programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,302 @@
+use crate::{
+    checks::check_token_source_account,
+    error::MarinadeError,
+    events::user::WithdrawStakeAccountEvent,
+    state::{stake_system::StakeSystem, validator_system::ValidatorSystem},
+    State,
+};
+
+use anchor_lang::solana_program::{native_token::LAMPORTS_PER_SOL, program::invoke_signed, stake};
+use anchor_lang::{
+    prelude::*,
+    solana_program::{program::invoke, stake::state::StakeAuthorize},
+};
+use anchor_spl::{
+    stake::{
+        deactivate_stake as solana_deactivate_stake, DeactivateStake as SolanaDeactivateStake,
+        Stake, StakeAccount,
+    },
+    token::{burn, Burn, Mint, Token, TokenAccount},
+};
+
+use crate::checks::check_stake_amount_and_validator;
+
+#[derive(Accounts)]
+pub struct WithdrawStakeAccount<'info> {
+    #[account(mut)]
+    pub state: Box<Account<'info, State>>,
+
+    #[account(mut)]
+    pub msol_mint: Box<Account<'info, Mint>>,
+
+    // Note: new stake account withdraw-auth (owner) & staker-auth will be owner of burn_msol_from
+    #[account(
+        mut,
+        token::mint = state.msol_mint
+    )]
+    pub burn_msol_from: Box<Account<'info, TokenAccount>>,
+    #[account(mut)]
+    pub burn_msol_authority: Signer<'info>,
+
+    /// CHECK: manual account processing
+    #[account(
+        mut,
+        address = state.validator_system.validator_list.account,
+        constraint = validator_list.data.borrow().as_ref().get(0..8)
+            == Some(ValidatorSystem::DISCRIMINATOR)
+            @ MarinadeError::InvalidValidatorListDiscriminator,
+    )]
+    pub validator_list: UncheckedAccount<'info>,
+
+    /// CHECK: manual account processing
+    #[account(
+        mut,
+        address = state.stake_system.stake_list.account,
+        constraint = stake_list.data.borrow().as_ref().get(0..8)
+            == Some(StakeSystem::DISCRIMINATOR)
+            @ MarinadeError::InvalidStakeListDiscriminator,
+    )]
+    pub stake_list: UncheckedAccount<'info>,
+    /// CHECK: PDA
+    #[account(
+        seeds = [
+            &state.key().to_bytes(),
+            StakeSystem::STAKE_WITHDRAW_SEED
+        ],
+        bump = state.stake_system.stake_withdraw_bump_seed
+    )]
+    pub stake_withdraw_authority: UncheckedAccount<'info>,
+    #[account(mut)]
+    pub stake_account: Box<Account<'info, StakeAccount>>,
+
+    #[account(
```

Comment:
Must be `init` also because it is a pattern I'm using for stake splits everywhere (and you need to refund split stake rent in case you are withdrawing whole stake)

Reply:
1) I'm trying to make it compatible with SPL-stake-pool

https://github.com/solana-labs/solana-program-library/blob/5965c9c30bbd21b7a61562ba39f6a422810e77c2/stake-pool/single-pool/src/processor.rs#L847

Looks like they require a stake-account, am I correct?

2) We always use the split_stake_account, so simpler and no need to manage 2 cases, and no refunds needed.

Reply:
ok, init added